caddyhttp/staticfiles/fileserver.go

```diff
@@ -80,17 +80,37 @@ func (fs FileServer) serveFile(w http.ResponseWriter, r *http.Request, name stri
 	}
 
 	// redirect to canonical path
-	url := r.URL.Path
+	u := r.URL.Path
 	if d.IsDir() {
 		// Ensure / at end of directory url
-		if !strings.HasSuffix(url, "/") {
-			Redirect(w, r, path.Base(url)+"/", http.StatusMovedPermanently)
+		if !strings.HasSuffix(u, "/") {
+			// Use the request's RequestURI since the path in u may have been
+			// trimmed by Caddy.
+			redirURL, err := url.ParseRequestURI(r.RequestURI)
+			if err != nil {
+				return http.StatusInternalServerError, err
+			}
+			if !strings.HasSuffix(redirURL.Path, "/") {
+				redirURL.Path = redirURL.Path + "/"
+			}
+			toURL := r.URL.ResolveReference(redirURL)
+
+			http.Redirect(w, r, toURL.String(), http.StatusMovedPermanently)
```

Comment:
Hmm, it's unfortunate that all this is required. Sorry about that. Since you've now worked with this a bit, do you have any recommendations for how we can make it easier/simpler?